<!--

Programmer: Victoria Albanese
Filename: camera_front.launch

Purpose: Launch the "front" Intel RealSense 435i camera
on the rigging so that its pointcloud can be visualized

-->

<launch>

    <!-- This launches the camera with the correct parameters -->
    <include file="$(find realsense2_camera)/launch/rs_rgbd.launch">
	<arg name="camera" value="camera_front" />
	<arg name="serial_no" value="843112073646" />
    </include>


    <!-- This transforms each resulting pointcloud into a global reference frame -->
    <node name="camera_front_tf" pkg="tf" type="static_transform_publisher"
        args="0.027 -0.388 0.40 0.01 -0.015 -2.0944 /world /camera_front_color_optical_frame 100"/>


    <!-- These nodelets limit the FOV of the camera (including filtering out the ground plane) -->
    <node pkg="nodelet" type="nodelet" name="cloud_front_filter_fov_depth" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_front/depth_registered/points_rotated"/>
        <remap from="~output" to="/camera_front/depth_registered/points_filtered_depth"/>
        <rosparam>
            filter_field_name: y
            filter_limit_min: -0.3
            filter_limit_max: 0.3
            filter_limit_negative: False
        </rosparam>
    </node>	
    <node pkg="nodelet" type="nodelet" name="cloud_front_filter_fov_sides" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_front/depth_registered/points_filtered_depth"/>
        <remap from="~output" to="/camera_front/depth_registered/points_filtered_sides"/>
        <rosparam>
            filter_field_name: x
            filter_limit_min: -0.3
            filter_limit_max:  0.3
            filter_limit_negative: False
        </rosparam>
    </node>	
    <node pkg="nodelet" type="nodelet" name="cloud_front_ground_filter" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_front/depth_registered/points_filtered_sides"/>
        <remap from="~output" to="/camera_front/depth_registered/points_filtered_ground_plane"/>
        <rosparam>
            filter_field_name: z
            filter_limit_min: 0.0075
            filter_limit_max: 0.3
            filter_limit_negative: False
        </rosparam>
    </node>	


    <!-- Statistical outlier removal for good measure -->
    <node pkg="nodelet" type="nodelet" name="cloud_front_outlier_removal_pass1" args="load pcl/StatisticalOutlierRemoval pcl_manager" output="screen"> 
        <remap from="~input"  to="/camera_front/depth_registered/points_filtered_ground_plane" /> 
        <remap from="~output" to="/camera_front/depth_registered/points_filtered" /> 
        <rosparam> 
               mean_k: 100 
               stddev: 0.9 
               negative: False 
        </rosparam> 
    </node> 

</launch>

